### PR TITLE
Decouple + implement validation/stream-creation using new approach

### PR DESCRIPTION
**Pull request by jrehwaldt**
This PR introduces a couple of interfaces and extension points, notably
* `KinesisInboundHandler` (implemented by `KinesisListenerProxy` and allows monitoring worker state)
* `KinesisOutboundStream` (kind of a per-stream based `AwsKinesisOutboundGateway`)
* `KinesisInboundHandlerPostProcessor` (allows to modify/decorate `KinesisInboundHandler`; think `Bean PostProcessor`)
* `KinesisOutboundStreamPostProcessor` (same for `KinesisOutboundStream`

Also, validation and stream creation is implemented using the post-processor approach.
Respective auto configuration is extracted into `KinesisCreateStreamAutoConfiguration` and `KinesisValidationAutoConfiguration`.

Features activation is controlled using properties `aws.kinesis.create-streams` (default: false) as well as `aws.kinesis.validate` (default: true).

### CODE REVIEW

File: src/main/kotlin/de/bringmeister/spring/aws/kinesis/KinesisListenerProxy.kt [hunk 1]
```diff
@@ -1,14 +1,19 @@
...
-) {
+    override val stream: String
+): KinesisInboundHandler {
 
-    fun invoke(data: Any?, metadata: Any?) {
-        method.invoke(bean, data, metadata)
+    override fun handleMessage(data: Any?, metadata: Any?) {
+        try {
+            method.invoke(bean, data, metadata)
+        } catch (ex: InvocationTargetException) {
```
Comment (jrehwaldt) [inline src/main/kotlin/de/bringmeister/spring/aws/kinesis/KinesisListenerProxy.kt]: This was a bug causing to bubble `InvocationTargetException` into the Worker thread.

**Comment by jrehwaldt** [comment]
As a side note: I made nearly all changes backwards compatible.

File: README.md [hunk 2]
```diff
@@ -85,23 +85,36 @@ Both tests will use the same Docker images in order to send and receive messages
...
 ```
 
-By default, `create-streams` will be turned-off. 
-So if you don't specify anything, no streams will be created.
+By default, `create-streams` is turned off and streams *must* be created externally.
+
+#### Validate data send and received
+
+By default, this starter validates all data send and received automatically if a bean of type `javax.validation.Validator` is found.
+Yu can disable this feature by setting `validate` flag to `false`:
```
Comment (andrejivanov) [inline README.md]: Typo: you

Comment (jrehwaldt) [reply README.md]: Thanks. Already fixed in #64. I'll see and wait if I have to backport it into this PR.

File: src/main/kotlin/de/bringmeister/spring/aws/kinesis/KinesisInboundHandler.kt [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    /** Indicates that the Worker is initialized and ready to send message. */
+    fun ready() { }
+
+    /**
+     * Called for each message.
+     *
+     * May be retried when an exception is thrown, except when throwing
+     * [UnrecoverableException].
+     */
+    fun handleMessage(data: Any?, metadata: Any?)
```
Comment (jrehwaldt) [inline src/main/kotlin/de/bringmeister/spring/aws/kinesis/KinesisInboundHandler.kt]: I‘d like to change this to

```kotlin
fun handleMessage(message: Message)

interface Message {
    fun data(): Any?
    fun meta(): Any?
    // extensible without breaking handler interface
    // e.g. by adding
    fun isRetry(): Boolean
}
```

What do you think, @andrejivanov?

Comment (andrejivanov) [reply src/main/kotlin/de/bringmeister/spring/aws/kinesis/KinesisInboundHandler.kt]: Yeah, that's good. "IsRetry" belongs more in the metadata, but is just an example.

**Comment by jrehwaldt** [comment]
I found a bug in the earlier implementation causing `ReflectionBasedRecordMapper` to not be able to infer the required types.

My last commit separates `RecordMapper`, which used to be a factory and mapper in one, into two parts:
* `RecordDeserializer` (to be in line with the method name `deserialize`)
* `RecordDeserializerFactory`

This also frees the code from constantly using reflection on each and every message in order to infer the required listener types, which could be a significant performance issue during load.

I planned to do this in another PR, but due to the issue mentioned above it had to be done as part of this one.

I'll wait for more input and don't merge yet.

**Comment by jrehwaldt** [comment]
Seems like our Coverall reports are flaky. Every second fails the check even though coverage effectively increases.

**Comment by jrehwaldt** [comment]
Yes, let’s wait for additional input.

I’ll try to improve the `RecordDeserializer` - `KinesisInboundHandler` interplay and see if I can move the post processor decoration into the `AwsInboundGateway.register` method. This way the decoration is consistently applied by both gateways, which it currently isn’t. But this has to wait until Monday. Nothing is anyway happening before.

Thanks for the review and let’s see what others think.

File: src/main/kotlin/de/bringmeister/spring/aws/kinesis/KinesisListenerProxyRecordDeserializerFactory.kt [hunk 4]
```diff
@@ -0,0 +1,32 @@
+package de.bringmeister.spring.aws.kinesis
+
+import com.fasterxml.jackson.databind.JavaType
+import com.fasterxml.jackson.databind.ObjectMapper
+import java.nio.charset.Charset
+
+class KinesisListenerProxyRecordDeserializerFactory(
```
[the remark this answers is not in the document]
Comment (jrehwaldt) [reply src/main/kotlin/de/bringmeister/spring/aws/kinesis/KinesisListenerProxyRecordDeserializerFactory.kt]: I renamed it to `ObjectMapperRecordDeserializerFactory` and moved the type decision into `KinesisInboundHandler.dataType|metaType()`. I feel this makes it cleaner and more reusable than what I had before.

File: src/main/kotlin/de/bringmeister/spring/aws/kinesis/KinesisInboundHandler.kt [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    /** Indicates that the Worker is initialized and ready to send message. */
+    fun ready() { }
+
+    /**
+     * Called for each message.
+     *
+     * May be retried when an exception is thrown, except when throwing
+     * [UnrecoverableException].
+     */
+    fun handleRecord(record: Record<*, *>, context: ExecutionContext)
```
[the remark this answers is not in the document]
Comment (jrehwaldt) [reply src/main/kotlin/de/bringmeister/spring/aws/kinesis/KinesisInboundHandler.kt]: Having `ExecutionContext` allows us to pass down additional information without breaking the public API `KinesisInboundHandler`. The `ExecutionContext` is under our control `AwsKinesisRecordProcessor` and unlikely to be implemented externally. On the other hand, library users might implement a variety of features using `KinesisInboundHandler`, such as auditing, metrics, ... Therefore, having an extra object passed down allows us to provide additional info in the future without breaking what users implement.

File: src/main/kotlin/de/bringmeister/spring/aws/kinesis/validation/KinesisValidationAutoConfiguration.kt [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+@ConditionalOnBean(Validator::class)
+@ConditionalOnProperty("aws.kinesis.validate", matchIfMissing = true)
+@AutoConfigureBefore(AwsKinesisAutoConfiguration::class)
+class KinesisValidationAutoConfiguration(
+    private val validator: Validator
+) {
+
+    @Bean
+    fun validatingPostProcessor() = ValidatingPostProcessor(validator)
+}
```
[the remark this answers is not in the document]
Comment (jrehwaldt) [reply src/main/kotlin/de/bringmeister/spring/aws/kinesis/validation/KinesisValidationAutoConfiguration.kt]: I prefer to have different concerns grouped each completely into their own package. What do others think? @andrejivanov @jmatusewicz

File: src/main/kotlin/de/bringmeister/spring/aws/kinesis/AwsKinesisOutboundStreamFactory.kt [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+    private val streams = ConcurrentHashMap<String, KinesisOutboundStream>()
+
+    override fun forStream(streamName: String): KinesisOutboundStream {
+        return streams.computeIfAbsent(streamName, this::createStream)
+    }
+
+    private fun createStream(streamName: String): KinesisOutboundStream {
+        val initial = AwsKinesisOutboundStream(streamName, requestFactory, clientProvider)
+        return postProcessors.fold(initial) { stream: KinesisOutboundStream, postProcessor ->
```
Comment (jmatusewicz) [inline src/main/kotlin/de/bringmeister/spring/aws/kinesis/AwsKinesisOutboundStreamFactory.kt]: Not sure if I understand this correctly but it looks like a decorator pattern for me. You are wrapping the initial stream with those post post processors, aren't you? Do we need to have control over the order of the wrapping layers?

Comment (jrehwaldt) [reply src/main/kotlin/de/bringmeister/spring/aws/kinesis/AwsKinesisOutboundStreamFactory.kt]: I have a test prepared showing how to adjust the ordering as part of another coming PR (metrics). Basically, you can use `@Priority` or `@Order` annotations on the `PostProcessor`.

**Pull request merged**